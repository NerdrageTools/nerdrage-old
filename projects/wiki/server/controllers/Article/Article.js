export default class Article {
  static defaults = {
    aliases: [],
    campaign: '',
    data: {},
    slug: '',
    tags: [],
    title: '',
  }

  static findOne = async (db, { campaign, slug }) => {
    const data = db.collection('articles').findOne({ campaign, slug })
    return new Article(db, data)
  }

  db = null

  constructor(db, data) {
    this.db = db
    this.data = { ...Article.defaults, ...data }
  }
}
