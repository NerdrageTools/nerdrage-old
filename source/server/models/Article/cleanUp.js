import cheerio from 'cheerio'
import parseHref from '@/utilities/parseHref'

export default function cleanUp(html) {
  const $ = cheerio.load(html || '', { decodeEntities: false, xmlMode: true })
  $('a').each((_, el) => {
    const $el = $(el)
    const href = $el.attr('href')

    if (!href.includes('://')) {
      const parsed = parseHref(href)
      const slug = $el.attr('slug') || parsed.slug
      const type = $el.attr('type') || parsed.type || 'article'
      const subdomain = $el.attr('subdomain') || parsed.subdomain
      $el.attr('href', `${subdomain}/${type}/${slug}`)
    }

    $el
      .removeClass('cross-campaign')
      .removeClass('external')
      .removeClass('missing')
      .removeClass($el.attr('type'))
      .removeAttr('slug')
      .removeAttr('subdomain')
      .removeAttr('target')
      .removeAttr('type')

    if (($el.attr('class') || '').trim() === '') {
      $el.removeAttr('class')
    }
  })
  $('include').each((_, el) => $(el).removeClass('noedit').html(''))

  return $.html().trim()
}
