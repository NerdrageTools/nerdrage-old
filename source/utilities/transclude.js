import Article from '@/server/models/Article'
import unique from '@/utilities/unique/unique'
import cheerio from 'cheerio'

export default function transclude(html) {
  const $ = cheerio.load(html, { xmlMode: true, decodeEntities: false })
  const links = []
  const missing = []

  return Promise.all($('include').map((_, includeEl) => {
    const $include = $(includeEl)
    $include.addClass('noedit')
    const includes = {}

    const transclusions = []
    const from = $include.attr('from')
    links.push(from)

    if (!includes[from]) {
      includes[from] = Article.findOne({ $or: [{ slug: from }, { aliases: from }] })
    }

    return includes[from].then((article) => {
      if (!article) {
        $include.html(`\n<div class="warning">Included Article '${from}' does not exist</div>\n`)
        links.push(from)
        missing.push(from)
        return Promise.resolve()
      }

      const $article = $.load(article.html, { xmlMode: true, decodeEntities: false })

      const sections = $include.attr('sections')
      if (sections === '*' || !sections) {
        transclusions.push($article.html())
      } else {
        sections.split(',').map(s => s.trim()).filter(Boolean).forEach((section) => {
          $article(`#${section}`).each((_, sectionEl) => {
            transclusions.push($article(sectionEl).html())
          })
        })
      }

      const firstPass = `\n${transclusions.map(line => `  ${line}`).join('\n')}\n`
      return transclude(firstPass).then((transcluded) => {
        $include.html(transcluded.html)
        links.push(...transcluded.links)
        missing.push(...transcluded.missing)
        return Promise.resolve(true)
      })
    })
  }).get())
    .then(() => ({
      html: $.html(),
      links: unique(links.filter(Boolean)).sort(),
      missing: unique(missing.filter(Boolean)).sort(),
    }))
}
