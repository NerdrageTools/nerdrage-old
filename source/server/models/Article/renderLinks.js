import url from 'url'
import cheerio from 'cheerio'
import Article from '@/server/models/Article'
import { extractUrlSlug } from '@/utilities/slugs'
import unique from '@/utilities/unique'

export default async function renderLinks(html, campaignId) {
  const $ = cheerio.load(html || '', { decodeEntities: false, xmlMode: true })
  const links = []

  $('a').each((_, element) => {
    const $link = $(element)
    const href = $link.attr('href') || ''
    const slug = extractUrlSlug(href)
    const parsedUrl = url.parse(href)

    const isInternal = !parsedUrl.hostname
    if (isInternal) {
      if (slug.match(/\.(gif|jpg|png)$/)) {
        $link.attr('href', `/media/${slug}`)
      } else {
        $link.attr('href', `/page/${slug}`)
      }
      links.push(slug)
    } else {
      $link.attr('target', '_new').addClass('external')
    }
  })

  const linked = await Article.find({
    $and: [
      { $or: [{ campaign: null }, { campaign: campaignId }] },
      { $or: [{ slug: { $in: links } }, { aliases: { $in: links } }] },
    ],
  })
    .select('slug aliases')
    .sort({ campaign: -1 })
    .exec()
  const foundLinks = (linked || []).reduce((all, { aliases, slug }) => (
    [...all, slug, ...aliases]
  ), [])

  const missing = links.filter(link => !foundLinks.includes(link))

  $('a').each((_, element) => {
    const $link = $(element)
    const href = $link.attr('href') || ''
    const slug = extractUrlSlug(href)
    if (missing.includes(slug)) {
      $link.addClass('missing')
    }
  })

  return Promise.resolve({
    html: $.html(),
    links: unique(links).sort(),
    missing: unique(missing).sort(),
  })
}
