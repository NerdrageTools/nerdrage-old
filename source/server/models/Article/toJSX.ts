import { decodeHTML, decodeXML, encodeHTML5 } from 'entities'
import { html_beautify as beautify } from 'js-beautify'
import { VOID_ELEMENTS } from '~/utilities/VOID_ELEMENTS'

const BEAUTIFY_OPTIONS = {
	end_with_newline: true,
	indent_char: ' ',
	indent_size: 2,
	preserve_newlines: false,
	wrap_line_length: 0,
}
const VOID_ELEMENT_CLOSING_TAGS = new RegExp(`<\/(${VOID_ELEMENTS.join('|')})>`, 'g')

export const toJSX = (html: string = '') => (
	([
		decodeHTML,
		decodeXML,
		(value: string) => value.replace(VOID_ELEMENT_CLOSING_TAGS, '')
	]).reduce((value, fn) => fn(value), html)
)
