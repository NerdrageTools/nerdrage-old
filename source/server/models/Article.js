import mongoose from 'mongoose'
import cleanUp from './Article/cleanUp'
import renderLinks from './Article/renderLinks'
import Slug from '@/server/models/Slug'
import transclude from '@/utilities/transclude'
import unique from '@/utilities/unique'

const { ObjectId: ObjectIdType } = mongoose.Schema.Types
const slugifyArray = array => unique(array.map(value => (
  value.toLowerCase().replace(/[^\w-_]/g, '-').replace(/-{2,}|^-|-$/g, '')
))).sort()

export const ArticleSchema = new mongoose.Schema({
  aliases: [Slug],
  campaign: { default: null, ref: 'Campaign', required: false, type: ObjectIdType, unique: false },
  createdBy: { ref: 'User', type: String },
  data: Object,
  html: { default: '', type: String },
  lastUpdatedBy: { ref: 'User', type: String },
  slug: { ...Slug, required: true, unique: false },
  tags: [Slug],
  title: { required: true, type: String },
}, {
  timestamps: true,
  toJSON: { virtuals: true },
  toObject: { virtuals: true },
  versionKey: 'version',
})

ArticleSchema.index({ campaign: 1, slug: 1 }, { unique: true })
ArticleSchema.pre('save', function (next) {
  this.aliases = slugifyArray(this.aliases)
  this.html = cleanUp(this.html)
  this.tags = slugifyArray(this.tags)

  mongoose.models.Article.updateMany(
    { _id: { $ne: this._id }, campaign: this.campaign },
    { $pull: { aliases: { $in: this.aliases } } }
  ).then(() => next())
})

ArticleSchema.methods.render = async function () {
  const links = await renderLinks(this.html, this.campaign)
  const includes = await transclude(links.html)

  return Promise.resolve({
    ...this.toJSON(),
    html: includes.html,
    links: unique([...links.links, ...includes.links]).sort(),
    missing: unique([...links.missing, ...includes.missing]).sort(),
  })
}

ArticleSchema.statics.locate = function (slug, campaignId) {
  return this.findOne({
    /* eslint-disable sort-keys */
    $and: [
      { $or: [{ campaign: campaignId }, { campaign: null }] },
      { $or: [{ aliases: slug }, { slug }] },
    ],
  /* eslint-ensable sort-keys */
  }).sort({ campaign: -1 })
}

ArticleSchema.pre('save', function () { this.compute() })

const Article = mongoose.model('Article', ArticleSchema)

export default Article
