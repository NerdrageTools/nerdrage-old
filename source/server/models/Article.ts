import {
	getModelForClass, index, modelOptions, pre, prop, Ref, ReturnModelType, Severity,
} from '@typegoose/typegoose'
import cheerio from 'cheerio'
import * as entities from 'entities'
import { html_beautify as beautify } from 'js-beautify'
import { Slug } from '~/server/models/Slug'
import { computeSearchKeys } from '~/utilities/computeSearchKeys'
import { createCampaignFilter } from '~/utilities/createCampaignFilter'
import { transclude } from '~/utilities/transclude'
import { unique } from '~/utilities/unique'
import { cleanUp } from '~/server/models/Article/cleanUp'
import { renderLinks } from '~/server/models/Article/renderLinks'
import { Campaign } from './Campaign'
import { User } from './User'

const slugifyArray = (array: string[]) => unique(array.map(value => (
	value.toLowerCase().replace(/[^\w-_]/g, '-').replace(/-{2,}|^-|-$/g, '')
))).sort()
const BEAUTIFY_OPTIONS = {
	end_with_newline: true,
	indent_char: ' ',
	indent_size: 2,
	preserve_newlines: false,
	wrap_line_length: 0,
}

export interface IArticleData {
	_id?: string,
	aliases?: string[],
	campaign: Ref<typeof Campaign>,
	createdAt?: Date,
	createdBy?: Ref<User>,
	html?: string,
	lastUpdatedBy?: Ref<User>,
	plainText?: string,
	secret?: boolean,
	slug: string,
	tags?: string[],
	template?: boolean,
	title: string,
	updatedAt?: Date,
	version?: number,
}

export interface IArticleLink {
	slug: string,
	subdomain: string,
	title: string,
}

export interface IArticleSearchResult {
	nameHits: number,
	preview: string[],
	secret: boolean,
	slug: string,
	subdomain: string,
	textHits: number,
	title: string,
}

@index({ campaign: 1, slug: 1 }, { unique: true })
@index({ plainText: 'text', title: 'text' })
@modelOptions({
	options: { allowMixed: Severity.ALLOW },
	schemaOptions: { id: false, timestamps: true, versionKey: 'version' },
})
@pre<Article>('save', async function (next) {
	this.aliases = slugifyArray(this.aliases)
	this.html = beautify(cleanUp(this.html), BEAUTIFY_OPTIONS)
	this.tags = slugifyArray(this.tags)

	const $ = cheerio.load(this.html, { decodeEntities: false, xmlMode: true })
	$('br').replaceWith('\r\n')
	// @ts-expect-error - .text() is perfectly valid on Root
	this.plainText = entities.decodeHTML($.text().replace(/[\r\n]/g, ' ').replace(/\s+/g, ' '))
	this.searchKeys = computeSearchKeys(this.plainText)

	await Articles.updateMany(
		{ _id: { $ne: this._id }, campaign: this.campaign },
		{ $pull: { aliases: { $in: this.aliases } } },
	)
	next()
})
export class Article implements IArticleData {
	_id?: string
	createdAt?: Date
	updatedAt?: Date
	version?: number

	@prop({ default: [], type: [String] })
	public aliases: string[] = []

	@prop({ default: null, ref: 'Campaign' })
	public campaign: Ref<typeof Campaign>

	@prop({ ref: 'User' })
	public createdBy?: Ref<User>

	@prop({ default: '' })
	public html: string = ''

	@prop({ ref: 'User' })
	public lastUpdatedBy?: Ref<User>

	@prop({ default: '' })
	public plainText: string = ''

	@prop({ default: false })
	public secret: boolean = false

	@prop({ ...Slug, required: true })
	public slug: string = ''

	@prop({ default: [], type: [String] })
	public tags: string[] = []

	@prop({ default: false })
	public template: boolean = false

	@prop({ required: true })
	public title: string = ''

	async render(campaign?: ICampaign): Promise<Article> {
		const campaignFilter = createCampaignFilter(campaign)
		const includes = await transclude(this.html, campaignFilter)
		const links = await renderLinks(includes.html, campaignFilter)
		const childArticles = await Articles.aggregate([
			{ $match: {
				$and: [campaignFilter, { tags: { $in: [this.slug, ...this.aliases] } }],
				template: false,
			} },
			{ $group: {
				_id: '$slug',
				id: { $first: '$_id' },
				slug: { $first: '$slug' },
				title: { $first: '$title' },
			} },
			{ $project: { _id: '$id', slug: '$slug', title: '$title' } },
		]).exec()

		const html = beautify(links.html, BEAUTIFY_OPTIONS)
		return Promise.resolve({ ...this.toJSON(), childArticles, html })
	}
}

export const Articles = getModelForClass(Article)
