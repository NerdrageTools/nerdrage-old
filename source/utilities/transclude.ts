import { DocumentType } from '@typegoose/typegoose'
import cheerio from 'cheerio'
import { ICampaignData } from '~/server/models'
import { Article, ArticleClass } from '~/server/models/Article'

export async function transclude(
	campaign: ICampaignData,
	html: string,
): Promise<string> {
	const $ = cheerio.load(html, { decodeEntities: false, xmlMode: true })
	const includes = $('include[from]')
	if (!includes.length) return html

	includes.map(async (_, el) => {
		const $el = $(el)
		const slug = $el.attr('from') ?? ''
		const sections = ($el.attr('sections') ?? '*').split(',').map(s => s.trim())
		const article = await Article.findByCampaignSlug(slug, campaign) as DocumentType<ArticleClass>

		if (!article) {
			$el.html(`\n<Warning>Transcluded article '${slug}' does not exist.</Warning>\n`)
		} else {
			const transclusions: string[] = []
			sections.forEach(async section => {
				if (section === '*') {
					transclusions.push(article.html ?? '')
				} else {
					transclusions.push(
						article?.sections.get(section)
							?? `<!-- Transcluded section '${section}' of article '${slug}' does not exist -->`,
					)
				}
			})
			$el.html(transclusions.join('\n'))
		}

		$el.attr('resolved', 'true')
	})

	return $.html()
}

// export async function transclude(
// 	campaign: ICampaignData,
// 	html: string,
// ): Promise<string> {
// 	const $ = cheerio.load(html, { decodeEntities: false, xmlMode: true })
// 	const includes = $('include').toArray()
// 	const slugs = unique<string>(
// 		includes
// 			.map(el => ($(el).attr('from') ?? '').trim().toLowerCase())
// 			.filter(Boolean),
// 	)
// 	const articles = await Articles.findByCampaignSlugs(slugs, campaign)

// 	includes.forEach(el => {
// 		const $el = $(el)
// 		$el.addClass('noedit')
// 		const slug = $el.attr('from')
// 		const article = articles.find(a => a.slug === slug)
// 		if (!article) {
// 			$el.html(`\n<Warning>Transcluded Article '${slug}' does not exist.</Warning>\n`)
// 		} else {
// 			const $article = cheerio.load(article.html, { decodeEntities: false, xmlMode: true })
// 			const sections = $el.attr('sections')
// 			if (sections === '*' || !sections) {
// 				$el.html($article.html())
// 			} else {
// 				let transclusion = ''
// 				sections.split(',').map(s => s.trim()).filter(Boolean).forEach(section => {
// 					$article(`#${section}`).each((_, sectionEl) => {
// 						transclusion += $article(sectionEl).html()
// 					})
// 				})
// 				$el.html(transclusion)
// 			}
// 		}
// 	})

// 	return html
// }

// 	// return Promise.all($('include').map((_, includeEl) => {
// 	// 	const $include = $(includeEl)
// 	// 	$include.addClass('noedit')
// 	// 	const includes = {}

// 	// 	const transclusions = []
// 	// 	const from = $include.attr('from')!
// 	// 	links.push(from)

// 	// 	if (!includes[from]) {
// 	// 		includes[from] = Articles.findOne({
// 	// 			$and: [
// 	// 				campaignFilter,
// 	// 				{ $or: [{ slug: from }, { aliases: from }] },
// 	// 			],
// 	// 		})
// 	// 	}

// 	// 	return includes[from].then(article => {
// 	// 		if (!article) {
// 	// 			$include.html(`\n<Warning>Transcluded Article '${from}' does not exist.</Warning>\n`)
// 	// 			links.push(from)
// 	// 			missing.push(from)
// 	// 			return Promise.resolve()
// 	// 		}

// 	// 		const $article = $.load(article.html, { decodeEntities: false, xmlMode: true })

// 	// 		const sections = $include.attr('sections')
// 	// 		if (sections === '*' || !sections) {
// 	// 			transclusions.push($article.html())
// 	// 		} else {
// 	// 			sections.split(',').map(s => s.trim()).filter(Boolean).forEach(section => {
// 	// 				$article(`#${section}`).each((_, sectionEl) => {
// 	// 					transclusions.push($article(sectionEl).html())
// 	// 				})
// 	// 			})
// 	// 		}

// 	// 		const firstPass = `\n${transclusions.map(line => `  ${line}`).join('\n')}\n`
// 	// 		return transclude(firstPass, campaignFilter).then(transcluded => {
// 	// 			$include.html(transcluded.html)
// 	// 			links.push(...transcluded.links)
// 	// 			missing.push(...transcluded.missing)
// 	// 			return Promise.resolve(true)
// 	// 		})
// 	// 	})
// 	// }).get())
// 	// 	.then(() => ({
// 	// 		html: $.html(),
// 	// 		links: unique(links.filter(Boolean)).sort(),
// 	// 		missing: unique(missing.filter(Boolean)).sort(),
// 	// 	}))
// }
