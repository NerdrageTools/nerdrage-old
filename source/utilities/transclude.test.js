import mongoose from 'mongoose'
import transclude from './transclude'
import Article from '@/server/models/Article'

describe('utilities/transclude', () => {
  beforeAll(async done => {
    await mongoose.connect('mongodb://localhost/test', {
      useCreateIndex: true,
      useNewUrlParser: true,
    })
    done()
  })

  it('transcludes', async done => {
    await Article.deleteMany({ slug: { $in: ['a', 'b', 'c'] } })
    await Article.create([
      /* eslint-disable sort-keys */
      { slug: 'a', title: 'Ant', html: 'Ant!' },
      { slug: 'b', title: 'Bug', html: 'Nope! <div id="simple">Bug!</div>' },
      { aliases: ['cat'], slug: 'c', title: 'Cat', html: `
        Nope!
        <div id="simple">Foo</div>
        <div id="nested">
          Bar
          <include sections="simple" from="b" />
        </div>
      ` },
      /* eslint-enable sort-keys */
    ])

    const html = `
      <h3>Title</h3>
      <include sections="*" from="a" />
      <include sections="simple" from="b" />
      <include sections="simple,nested" from="c" />
      <include sections="simple,nested" from="cat" />
      <include sections="*" from="d" />
      <include sections="missing,missing-2" from="e" />
    `

    const transcluded = await transclude(html)
    expect(transcluded.links).toEqual(['a', 'b', 'c', 'cat', 'd', 'e'])
    expect(transcluded.missing).toEqual(['d', 'e'])
    expect(transcluded.html).toMatchSnapshot()

    done()
  })

  afterAll(async done => {
    await mongoose.disconnect()
    done()
  })
})
