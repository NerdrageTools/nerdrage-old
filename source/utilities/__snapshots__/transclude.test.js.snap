// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utilities/transclude transcludes 1`] = `
"
      <h3>Title</h3>
      <include sections=\\"*\\" from=\\"a\\" class=\\"noedit\\">
  Ant!

</include>
      <include sections=\\"simple\\" from=\\"b\\" class=\\"noedit\\">
  Bug!
</include>
      <include sections=\\"simple,nested\\" from=\\"c\\" class=\\"noedit\\">
  Foo
   Bar
  <include sections=\\"simple\\" from=\\"b\\" class=\\"noedit\\">
  Bug!
</include>

</include>
      <include sections=\\"simple,nested\\" from=\\"cat\\" class=\\"noedit\\">
  Foo
   Bar
  <include sections=\\"simple\\" from=\\"b\\" class=\\"noedit\\">
  Bug!
</include>

</include>
      <include sections=\\"*\\" from=\\"d\\" class=\\"noedit\\">
<div class=\\"warning\\">Included Article 'd' does not exist</div>
</include>
      <include sections=\\"missing,missing-2\\" from=\\"e\\" class=\\"noedit\\">
<div class=\\"warning\\">Included Article 'e' does not exist</div>
</include>
    "
`;
